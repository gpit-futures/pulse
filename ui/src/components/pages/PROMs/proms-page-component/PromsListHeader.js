import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import PTButton from '../../../ui-elements/PTButton/PTButton';

const VIEW_CONTENT = 'viewContent';

export default class PromsListHeader extends PureComponent {
  static propTypes = {
    onFilterChange: PropTypes.func.isRequired,
    panelTitle: PropTypes.string.isRequired,
  };

  state = {
    isFilterInputVisible: false,
    selected: '',
    openedPanel: '',
    isFilterOpen: false,
  };

  componentWillMount() {
    document.addEventListener('click', this.handleClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false);
  }

  toggleFilterInputVisibility = /* istanbul ignore next */ () => this.setState(prevState => ({ isFilterInputVisible: !prevState.isFilterInputVisible, isFilterOpen: !prevState.isFilterOpen }),
    () => !this.state.isFilterInputVisible && this.props.onFilterChange({ target: { value: '' } })
  );

  handleClick = /* istanbul ignore next */ (e) => {
    if (!this.node.contains(e.target)) {
      this.setState({ openedPanel: '' });
    }
  };

  handleMouseDown = (name) => {
    this.setState((prevState) => {
      if (prevState.openedPanel !== name) {
        return ({ openedPanel: name })
      }
      return ({ openedPanel: '' })
    })
  };

  handleClickFilterFocus = () => this.setState({ openedPanel: '' });

  render() {
    const { isFilterInputVisible, openedPanel } = this.state;
    const { onFilterChange, panelTitle, isBtnExpandVisible, onExpand, name, currentPanel, activeView, toggleViewVisibility } = this.props;

    return (
      <div className="panel-heading" ref={node => this.node = node}>
        <div className="control-group right">
          <PTButton className="btn btn-success btn-inverse btn-filter" onClick={this.toggleFilterInputVisibility}>
            <i className="btn-icon fa fa-filter" />
          </PTButton>
        </div>
        <div className="control-group right" >
          { isBtnExpandVisible ? <PTButton className="btn btn-success btn-inverse btn-square hidden-xs hidden-sm btn-expand-panel" onClick={() => onExpand(name, currentPanel)}>
            <i className="btn-icon fa fa-expand" />
            <i className="btn-icon fa fa-compress" />
          </PTButton> : null }
          <div className={classNames('dropdown', { 'open': openedPanel === VIEW_CONTENT })}>
            <PTButton className="btn btn-success btn-inverse btn-dropdown-toggle btn-table" onClick={() => this.handleMouseDown(VIEW_CONTENT)}>
              {activeView === 'tableNews' ? <i className="btn-icon fa fa-table" /> : null }
              {activeView === 'chartNews' ? <i className="btn-icon fa fa-area-chart" /> : null }
            </PTButton>
            <div className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size">
              <div className="heading">TABLES</div>
              <div className="dropdown-menu-list">
                <div className={classNames('dropdown-menu-item', { 'active': activeView === 'tableNews' })} onClick={() => toggleViewVisibility('tableNews')}>
                  <i className="dropdown-menu-item-icon fa fa-table" />
                  <span className="dropdown-menu-item-text">Proms</span>
                </div>
              </div>
              <div className="heading">CHARTS</div>
              <div className="dropdown-menu-wrap-list">
                <div className="dropdown-menu-list">
                  <div className={classNames('dropdown-menu-item', { 'active': activeView === 'chartNews' })} onClick={() => toggleViewVisibility('chartNews')}>
                    <i className="dropdown-menu-item-icon fa fa-area-chart" />
                    <span className="dropdown-menu-item-text">Proms</span></div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <h3 className="panel-title">{panelTitle}</h3>
        { (isFilterInputVisible) ? <div className="panel-filter">
          <div className="inner-addon addon-left">
            <div className="addon">
              <i className="fa fa-filter" />
            </div>
            <input className="form-control" placeholder="Filter..." onChange={onFilterChange} onClick={() => this.handleClickFilterFocus()} autoFocus />
          </div>
        </div> : null}
      </div>
    )
  }
}
