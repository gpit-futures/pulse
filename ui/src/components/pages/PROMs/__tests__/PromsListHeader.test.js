import React from 'react';
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15';

import PromsListHeader from '../proms-page-component/PromsListHeader';

configure({ adapter: new Adapter() });

const testProps = {
  onExpand: () => {},
  onFilterChange: () => {},
  toggleViewVisibility: () => {},
  name: 'promsMain',
  panelTitle: 'Test Panel Title',
  activeView: 'tableNews',
  currentPanel: 'promsMain',
  isBtnExpandVisible: true,
  isBtnTableVisible: true,
};

describe('Component <PromsListHeader />', () => {
  it('should renders with all props correctly', () => {
    const component = shallow(
      <PromsListHeader
        onExpand={testProps.onExpand}
        onFilterChange={testProps.onFilterChange}
        toggleViewVisibility={testProps.toggleViewVisibility}
        name={testProps.name}
        currentPanel={testProps.currentPanel}
        panelTitle={testProps.panelTitle}
        isBtnExpandVisible
        isBtnTableVisible
        activeView={testProps.activeView}
      />);

    expect(component.find('PTButton')).toHaveLength(3);
    expect(component.find('.fa-expand')).toHaveLength(1);
    expect(component.find('.fa-filter')).toHaveLength(1);
    expect(component.find('.fa-compress')).toHaveLength(1);
    expect(component.find('.btn-table')).toHaveLength(1);
    expect(component.find('.fa-area-chart')).toHaveLength(1);

    expect(component.find('.heading').at(0).text()).toEqual('TABLES');
    expect(component.find('.heading').at(1).text()).toEqual('CHARTS');

    expect(component.find('.dropdown-menu-item-text').at(0).text()).toEqual('Proms');
    expect(component.find('.dropdown-menu-item-text').at(1).text()).toEqual('Proms');

    expect(component.find('.panel-title')).toHaveLength(1);
    expect(component.find('.panel-title').text()).toEqual(testProps.panelTitle);

    expect(component.find('input')).toHaveLength(0);
    expect(component).toMatchSnapshot();

    // Testing panel filter
    expect(component.find('.panel-filter')).toHaveLength(0);
    expect(component.state().isFilterInputVisible).toEqual(false);
    expect(component.state().isFilterOpen).toEqual(false);
    component.find('.btn-filter').simulate('click');
    expect(component.state().isFilterInputVisible).toEqual(true);
    expect(component.state().isFilterOpen).toEqual(true);
    expect(component.find('.panel-filter')).toHaveLength(1);
    expect(component).toMatchSnapshot();

  });

  it('should renders without some buttons correctly', () => {
    const component = shallow(
      <PromsListHeader
        toggleViewVisibility={testProps.toggleViewVisibility}
        activeView={testProps.activeView}
        nameShouldInclude={testProps.nameShouldInclude}
        onExpand={testProps.onExpand}
        onFilterChange={testProps.onFilterChange}
        name={testProps.name}
        currentPanel={testProps.currentPanel}
        panelTitle={testProps.panelTitle}
        isBtnExpandVisible={false}
        isBtnTableVisible={false}
      />);

    expect(component.find('.control-group').children()).toHaveLength(2);
    expect(component.find('PTButton')).toHaveLength(2);

    expect(component).toMatchSnapshot();

    expect(component.instance().props.isBtnExpandVisible).toEqual(false);
    expect(component.instance().props.isBtnTableVisible).toEqual(false);
  });

  it('should renders correctly and testing component methods', () => {
    const component = shallow(
      <PromsListHeader
        onExpand={testProps.onExpand}
        onFilterChange={testProps.onFilterChange}
        toggleViewVisibility={testProps.toggleViewVisibility}
        name={testProps.name}
        currentPanel={testProps.currentPanel}
        panelTitle={testProps.panelTitle}
        isBtnExpandVisible
        isBtnTableVisible
        activeView={testProps.activeView}
        nameShouldInclude={testProps.nameShouldInclude}
      />);

    // Testing toggleFilterInputVisibility methods
    expect(component.state().isFilterOpen).toEqual(false);
    expect(component.state().isFilterInputVisible).toEqual(false);
    expect(component.contains('.active')).toEqual(false);
    component.instance().toggleFilterInputVisibility();
    expect(component.state().isFilterOpen).toEqual(true);
    expect(component.state().isFilterInputVisible).toEqual(true);

    // Testing dropdown-menu-item click
    component.find('.dropdown-menu-item').at(0).simulate('click');
    component.find('.dropdown-menu-item').at(1).simulate('click');

    // Testing btn-dropdown-toggle click
    component.find('.btn-dropdown-toggle').at(0).simulate('click');

    // Testing btn-expand-panel click
    component.find('.btn-expand-panel').at(0).simulate('click');

    // Testing handleClickFilterFocus methods
    component.setState({ openedPanel: 'test' });
    component.instance().handleClickFilterFocus();
    expect(component.state().openedPanel).toEqual('');
    component.setState({ openedPanel: 'test' });
    component.find('.form-control').at(0).simulate('click');
    expect(component.state().openedPanel).toEqual('');

    // Testing handleMouseDown methods
    expect(component.state().openedPanel).toEqual('');
    component.instance().handleMouseDown('viewContent');
    expect(component.state().openedPanel).toEqual('viewContent');
    component.instance().handleMouseDown('viewContent');
    expect(component.state().openedPanel).toEqual('');

    component.setProps({ activeView: 'chartNews'});

    expect(component).toMatchSnapshot();

    // Testing componentWillUnmount method
    component.unmount();
    expect(component).toMatchSnapshot();
  });
});
