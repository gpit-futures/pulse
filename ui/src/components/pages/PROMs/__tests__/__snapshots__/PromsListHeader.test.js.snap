// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PromsListHeader /> should renders correctly and testing component methods 1`] = `
<div
  className="panel-heading"
>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-filter"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-filter"
      />
    </PTButton>
  </div>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-square hidden-xs hidden-sm btn-expand-panel"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-expand"
      />
      <i
        className="btn-icon fa fa-compress"
      />
    </PTButton>
    <div
      className="dropdown"
    >
      <PTButton
        className="btn btn-success btn-inverse btn-dropdown-toggle btn-table"
        onClick={[Function]}
      >
        <i
          className="btn-icon fa fa-area-chart"
        />
      </PTButton>
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size"
      >
        <div
          className="heading"
        >
          TABLES
        </div>
        <div
          className="dropdown-menu-list"
        >
          <div
            className="dropdown-menu-item"
            onClick={[Function]}
          >
            <i
              className="dropdown-menu-item-icon fa fa-table"
            />
            <span
              className="dropdown-menu-item-text"
            >
              Proms
            </span>
          </div>
        </div>
        <div
          className="heading"
        >
          CHARTS
        </div>
        <div
          className="dropdown-menu-wrap-list"
        >
          <div
            className="dropdown-menu-list"
          >
            <div
              className="dropdown-menu-item active"
              onClick={[Function]}
            >
              <i
                className="dropdown-menu-item-icon fa fa-area-chart"
              />
              <span
                className="dropdown-menu-item-text"
              >
                Proms
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    className="panel-title"
  >
    Test Panel Title
  </h3>
  <div
    className="panel-filter"
  >
    <div
      className="inner-addon addon-left"
    >
      <div
        className="addon"
      >
        <i
          className="fa fa-filter"
        />
      </div>
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Filter..."
      />
    </div>
  </div>
</div>
`;

exports[`Component <PromsListHeader /> should renders correctly and testing component methods 2`] = `
<div
  className="panel-heading"
>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-filter"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-filter"
      />
    </PTButton>
  </div>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-square hidden-xs hidden-sm btn-expand-panel"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-expand"
      />
      <i
        className="btn-icon fa fa-compress"
      />
    </PTButton>
    <div
      className="dropdown"
    >
      <PTButton
        className="btn btn-success btn-inverse btn-dropdown-toggle btn-table"
        onClick={[Function]}
      >
        <i
          className="btn-icon fa fa-area-chart"
        />
      </PTButton>
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size"
      >
        <div
          className="heading"
        >
          TABLES
        </div>
        <div
          className="dropdown-menu-list"
        >
          <div
            className="dropdown-menu-item"
            onClick={[Function]}
          >
            <i
              className="dropdown-menu-item-icon fa fa-table"
            />
            <span
              className="dropdown-menu-item-text"
            >
              Proms
            </span>
          </div>
        </div>
        <div
          className="heading"
        >
          CHARTS
        </div>
        <div
          className="dropdown-menu-wrap-list"
        >
          <div
            className="dropdown-menu-list"
          >
            <div
              className="dropdown-menu-item active"
              onClick={[Function]}
            >
              <i
                className="dropdown-menu-item-icon fa fa-area-chart"
              />
              <span
                className="dropdown-menu-item-text"
              >
                Proms
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    className="panel-title"
  >
    Test Panel Title
  </h3>
  <div
    className="panel-filter"
  >
    <div
      className="inner-addon addon-left"
    >
      <div
        className="addon"
      >
        <i
          className="fa fa-filter"
        />
      </div>
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Filter..."
      />
    </div>
  </div>
</div>
`;

exports[`Component <PromsListHeader /> should renders with all props correctly 1`] = `
<div
  className="panel-heading"
>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-filter"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-filter"
      />
    </PTButton>
  </div>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-square hidden-xs hidden-sm btn-expand-panel"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-expand"
      />
      <i
        className="btn-icon fa fa-compress"
      />
    </PTButton>
    <div
      className="dropdown"
    >
      <PTButton
        className="btn btn-success btn-inverse btn-dropdown-toggle btn-table"
        onClick={[Function]}
      >
        <i
          className="btn-icon fa fa-table"
        />
      </PTButton>
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size"
      >
        <div
          className="heading"
        >
          TABLES
        </div>
        <div
          className="dropdown-menu-list"
        >
          <div
            className="dropdown-menu-item active"
            onClick={[Function]}
          >
            <i
              className="dropdown-menu-item-icon fa fa-table"
            />
            <span
              className="dropdown-menu-item-text"
            >
              Proms
            </span>
          </div>
        </div>
        <div
          className="heading"
        >
          CHARTS
        </div>
        <div
          className="dropdown-menu-wrap-list"
        >
          <div
            className="dropdown-menu-list"
          >
            <div
              className="dropdown-menu-item"
              onClick={[Function]}
            >
              <i
                className="dropdown-menu-item-icon fa fa-area-chart"
              />
              <span
                className="dropdown-menu-item-text"
              >
                Proms
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    className="panel-title"
  >
    Test Panel Title
  </h3>
</div>
`;

exports[`Component <PromsListHeader /> should renders with all props correctly 2`] = `
<div
  className="panel-heading"
>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-filter"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-filter"
      />
    </PTButton>
  </div>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-square hidden-xs hidden-sm btn-expand-panel"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-expand"
      />
      <i
        className="btn-icon fa fa-compress"
      />
    </PTButton>
    <div
      className="dropdown"
    >
      <PTButton
        className="btn btn-success btn-inverse btn-dropdown-toggle btn-table"
        onClick={[Function]}
      >
        <i
          className="btn-icon fa fa-table"
        />
      </PTButton>
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size"
      >
        <div
          className="heading"
        >
          TABLES
        </div>
        <div
          className="dropdown-menu-list"
        >
          <div
            className="dropdown-menu-item active"
            onClick={[Function]}
          >
            <i
              className="dropdown-menu-item-icon fa fa-table"
            />
            <span
              className="dropdown-menu-item-text"
            >
              Proms
            </span>
          </div>
        </div>
        <div
          className="heading"
        >
          CHARTS
        </div>
        <div
          className="dropdown-menu-wrap-list"
        >
          <div
            className="dropdown-menu-list"
          >
            <div
              className="dropdown-menu-item"
              onClick={[Function]}
            >
              <i
                className="dropdown-menu-item-icon fa fa-area-chart"
              />
              <span
                className="dropdown-menu-item-text"
              >
                Proms
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    className="panel-title"
  >
    Test Panel Title
  </h3>
  <div
    className="panel-filter"
  >
    <div
      className="inner-addon addon-left"
    >
      <div
        className="addon"
      >
        <i
          className="fa fa-filter"
        />
      </div>
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Filter..."
      />
    </div>
  </div>
</div>
`;

exports[`Component <PromsListHeader /> should renders without some buttons correctly 1`] = `
<div
  className="panel-heading"
>
  <div
    className="control-group right"
  >
    <PTButton
      className="btn btn-success btn-inverse btn-filter"
      onClick={[Function]}
    >
      <i
        className="btn-icon fa fa-filter"
      />
    </PTButton>
  </div>
  <div
    className="control-group right"
  >
    <div
      className="dropdown"
    >
      <PTButton
        className="btn btn-success btn-inverse btn-dropdown-toggle btn-table"
        onClick={[Function]}
      >
        <i
          className="btn-icon fa fa-table"
        />
      </PTButton>
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-small-size"
      >
        <div
          className="heading"
        >
          TABLES
        </div>
        <div
          className="dropdown-menu-list"
        >
          <div
            className="dropdown-menu-item active"
            onClick={[Function]}
          >
            <i
              className="dropdown-menu-item-icon fa fa-table"
            />
            <span
              className="dropdown-menu-item-text"
            >
              Proms
            </span>
          </div>
        </div>
        <div
          className="heading"
        >
          CHARTS
        </div>
        <div
          className="dropdown-menu-wrap-list"
        >
          <div
            className="dropdown-menu-list"
          >
            <div
              className="dropdown-menu-item"
              onClick={[Function]}
            >
              <i
                className="dropdown-menu-item-icon fa fa-area-chart"
              />
              <span
                className="dropdown-menu-item-text"
              >
                Proms
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    className="panel-title"
  >
    Test Panel Title
  </h3>
</div>
`;
